### DISCUSSION TITLE
Unable to install BthPS3

### DISCUSSION THREAD

**Discussion by Necrow51** [nefarius/BthPS3#56] [Support]
Hello, I wanted to do an uninstallation and reinstallation but impossible
![187499201-83e675d9-80a8-43ed-b1e3-04fe1581c8d5](https://user-images.githubusercontent.com/88400526/187618478-701305dc-00a3-4333-a21f-76a7ff855e34.png)
Yet everything is well deleted in my drivers
![187499786-216dd10f-9a6a-4c15-9e8d-f9585bf21034](https://user-images.githubusercontent.com/88400526/187618506-93e4768d-85ec-4e5e-b0cb-675bc654ba63.jpg)
So I'm completely stuck :(

**Comment by nefarius** [comment]
The install check has nothing to do whether the drivers exist on your system or not, it checks if any previous MSI is found, so your registry is broken, Google "how to force uninstall broken MSI" or similar, good luck!

**Reply by CapMacMark** [reply]
Same issue here, still can find a good solution with force uninstall broken MSi. I tried to find it in registry using Computer\HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Installer\UserData\S-1-5-18\Products\ but i dont know the real product key of BthPS3 since GUID can be change in software SUPERORCA. Please give at least a hint how to solve it.

**Reply by Soupdragon1** [reply]
Same here too. The drivers are still there you can see it in Windows\System32\Drivers\BthPS3PSM.sys which lists it as Nefarius Software Solutions

**Reply by nefarius** [reply]
[Found and documented the solution](https://vigem.org/projects/BthPS3/Frequently-Asked-Questions/#how-to-fix-previous-version-found-on-reinstall)!

**Comment by Necrow51** [comment]
I can't find a solution that works, I tried to search for "nefarius" in my regedit but noting...and I don't have any file in my system called MSI

**Reply by Trashv1s** [reply]
did u find a solution?

**Reply by Soupdragon1** [reply]
Search for BTHPS3

**Reply by nefarius** [reply]
[Found and documented the solution](https://vigem.org/projects/BthPS3/Frequently-Asked-Questions/#how-to-fix-previous-version-found-on-reinstall)!

**Comment by r-roker** [comment]
I have the exact same problem on Windows 11. Uninstalled BthPS3 through the uninstaller and got this error when I tried to reinstall the driver. I tried messing around with the registry but could not find a solution. When I search for "BthPS3" in the registry, I find 2 entries but those cannot be deleted.

For now I am using my controllers through cable instead of Bluetooth. Any help or solution will be appreciated. Thanks in advance.

**Reply by Trashv1s** [reply]
have a solution?

**Reply by Soupdragon1** [reply]
No they can't be deleted, TrustedInstaller permissions only afaik

**Reply by nefarius** [reply]
[Found and documented the solution](https://vigem.org/projects/BthPS3/Frequently-Asked-Questions/#how-to-fix-previous-version-found-on-reinstall)!

**Comment by khorezm0** [comment]
**Solved, check out the comment**

After uninstalling (and then installing again) i got same error.

I did some investigation, and found out that in code there is check for existing "\\.\Device\BthPS3PSMControl" or just "\\.\BthPS3PSMControl" files. Those file is somehow end up left on my pc (you can check it with winobj.exe). So i guess if i find out how to delete them, then everything is going to be ok.

![image](https://user-images.githubusercontent.com/44477986/194536846-e4748167-b2d6-4f62-bed3-d1827781dc68.png)

**Reply by khorezm0** [reply]
ok. I  just deleted C:\Windows\System32\drivers\BthPS3PSM.sys (I cannot recommend you to do so, its dangerous to delete files in windows folder, but if you know what are you doing you can use Unlocker app to do so). And since then i able to intstall bthps3 again.

Edit: do not forget to clean registry at first, to delete all mentions of bthps3

**Reply by nefarius** [reply]
I can not stress enough how bad of an idea it is to recommend anybody without the proper knowledge to remove files from the system directories. Or to "clean the registry". I'll add BthPS3 emergency removal to Legacinator when I got the time.

**Reply by r-roker** [reply]
> ok. I just deleted C:\Windows\System32\drivers\BthPS3PSM.sys (I cannot recommend you to do so, its dangerous to delete files in windows folder, but if you know what are you doing you can use Unlocker app to do so). And since then i able to intstall bthps3 again.
> 
> Edit: do not forget to clean registry at first, to delete all mentions of bthps3

Please do not do this unless you exactly know what you are doing. I tried it and it wiped my bluetooth stack which could not be restored despite my best attempts including reinstallation of bluetooth drivers.

My Bluetooth stack got restored only after complete wipe and reinstallation of Windows.

Also the suggested Unlocker app comes bundled with adware which gets picked up by AV software.

**Reply by Trashv1s** [reply]
> I can not stress enough how bad of an idea it is to recommend anybody without the proper knowledge to remove files from the system directories. Or to "clean the registry". I'll add BthPS3 emergency removal to Legacinator when I got the time.

um can u help remove this? littlebit annoying cant reinstal this bthps3 again. i think many user have this problem and he wont to comment. pls

**Reply by Trashv1s** [reply]
> I can not stress enough how bad of an idea it is to recommend anybody without the proper knowledge to remove files from the system directories. Or to "clean the registry". I'll add BthPS3 emergency removal to Legacinator when I got the time.

i have waited 43 days since ur comment said will help with adding bthps3 emergency removal XD

**Reply by nefarius** [reply]
Repeatedly asking won't magically grant me more time or interest in this topic. How about you folks find a solution for a change? In open source spirit instead of always waiting that I jump in. I got enough other things to do.

**Reply by mstrzelec** [reply]
this happened to me after removing the bth installer from 'uninstall programs' instead of using the same MSI... not sure why, but removing the one (and only one) driver (BthPS3PSM.sys - you need permissions / ownership over it or use system account) and reboot solved it - i.e. I was able to reinstall.

Overall all my issues came to be from the fact I was not focused on the initial instruction, which is perfect if you do exactly what they say. 

Anyway thanks for all the help everyone <3

**Reply by Soupdragon1** [reply]
It seems to happen when you select uninstall from Add/Remove Programs, it removes itself from the list but leaves the drivers etc behind then since theres no way of removing drivers manually it won't let you install any new ones. Obviously you can reinstall windows from scratch but I'd rather avoid that if at all possible

**Reply by nefarius** [reply]
[Found and documented the solution](https://vigem.org/projects/BthPS3/Frequently-Asked-Questions/#how-to-fix-previous-version-found-on-reinstall)!

**Comment by peamy** [comment]
I'm having the same issue as @Necrow51.
I used to have everything installed on windows 10. After doing a Windows full reset (lose files and programs), I tried to install BthPS3 again, but it said it was already installed. Seems the clean install for windows left some things behind. However, I'm not able to remove everything.
Also checked the drivers like Necrow51 did and couldnt find any nefarius software.

Any tips on how to (un)install it?

**Reply by Trashv1s** [reply]
do u find a solution?

**Reply by peamy** [reply]
Unfortunately not. I fell back on SCP toolkit. It still suits my needs luckily.

**Reply by Trashv1s** [reply]
> Unfortunately not. I fell back on SCP toolkit. It still suits my needs luckily.

yeah we're same XD

**Reply by Soupdragon1** [reply]
`it said it was already installed`
sounds like you installed windows over an existing install which will simply update/upgrade it, you need to wipe the drive or remove the partition on the drive first

**Comment by Timoteee** [comment]
I developed my solution : 

How to replicate the bug
 1. Install Bthps3
 2. Reset Windows or Upgrade from an older version with bthps3 installed to a higher version without uninstallation of bthps3
 3. Try to reinstall bthps3
 4. The program would give the error as seen in the op's error image stating that the driver is already installed

Upon seeing the BthPS3PSM.sys located in my system32 drivers folder I came under the understanding that there were remnants of the bthps3 driver on my system. 

After analyzing the lower filter driver code (https://github.com/ViGEm/BthPS3/blob/master/BthPS3PSM/BthPS3PSM.inf)
looking at the delete function like 69-71. 
`[BthPS3PSM_DelReg]
;Delete BthPS3PSM from LowerFilters
HKLM, System\CurrentControlSet\Control\Class\{e0cbf06c-cd8b-4647-bb8a-263b43f0f974}, LowerFilters, 0x00018002, BthPS3PSM
`
It appeared I needed to perform a manual delete of any remaining bthps3psm registry code, since I didn't run the uninstaller. Following the location as mentioned in the code  :

> HKLM, System\CurrentControlSet\Control\Class\{e0cbf06c-cd8b-4647-bb8a-263b43f0f974}

**WARNING: IF YOU AREN'T PROFICIENT ENOUGH TO USE REGISTRY EDIT I RECOMMEND NOT PERFORMING THIS SOLUTION ALTHOUGH EASY TO DO, DO IT AT YOUR OWN RISK.**
**REMEMBER TO ALWAYS MAKE A BACKUP OF YOUR REGISTRY BEFORE MAKING EDITS**
**I AM NOT HOLDING MYSELF RESPONSIBLE FOR ANY MISTAKES THAT MAY HAPPEN DUE TO HUMAN ERROR**
**I'D RECOMMEND ALSO HAVING A WINDOWS INSTALLATION NEARBY IF ANYTHING DOES HAPPEN**

All we need to do is : 
1. run Regedit as admin
2. Go to this address 

> Computer\HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\Class\{e0cbf06c-cd8b-4647-bb8a-263b43f0f974}

3. Delete the line "BthPS3PSM" from the value of the LowerFilters 
![image](https://user-images.githubusercontent.com/86269369/205994622-bc632b94-e6fe-46fb-87cd-0d746945d8c7.png)
4.  Press ok and save
5. Once done editing the registry, we need to perform a delete  of the BthPS3PSM.sys file in the system32 drivers folder.
6.  Following this location 

> C:\Windows\System32\drivers 
we see a file named "BthPS3PSM.sys"

![image](https://user-images.githubusercontent.com/86269369/205998079-edeae09a-dafb-4cf6-955d-f9463969d799.png)
7. Right-click on it> Properties > Security Tab > Advanced > Select Changes (Top Center next to TrustedInstaller) > And enter the name of the current user and apply changes 
![image](https://user-images.githubusercontent.com/86269369/205995789-1f618e8c-3acb-4f62-a277-02a5d7c96c98.png)
8. Give all users full access to the file and exit the permissions editor.
9. Open a command prompt with admin privileges 
10. Type the following line into it and save `del C:\Windows\System32\drivers\BthPS3PSM.sys`
11. Reinstall bthps3 
12. Reboot 

I redid this two times, and in one case i didn't delete the registry file and just deleted the sys file and which resulted in my Bluetooth function and stack unusable meaning that my computer couldn't communicate via Bluetooth. Hope this helped

**Reply by skruffied** [reply]
I did all the steps and it just said that the access is denied :/

**Reply by Timoteee** [reply]
If its just the file you need to delete, then you're half way there, just figure out how to give yourself access to the file, with full access to it. I'm not home at the moment but I'll help when I'm back

**Reply by nefarius** [reply]
Try [this Beta release of Legacinator](https://github.com/nefarius/Legacinator/releases/tag/v1.1.42) please.

**Reply by nefarius** [reply]
Just released [the fixed version](https://github.com/nefarius/Legacinator/releases/tag/v1.2.48).

**Reply by daptheHuman** [reply]
> Just released [the fixed version](https://github.com/nefarius/Legacinator/releases/tag/v1.2.48).

legacinator still cannot detect the bthps3

**Reply by nefarius** [reply]
Sure does on my end, so... 🤷

**Reply by Soupdragon1** [reply]
Interestingly theres nothing BthPs3 listed at all under that registry key in mine, the driver.sys file is definitely there though. Even though I granted myself permission it won't delete says its in use...

Legacinator detects legacy drivers but can't remove them. Hitting the button greys the window momentarily then it returns to the default window with the button saying BthPS3 drivers found... press it as many times as you like it makes no difference

![image](https://user-images.githubusercontent.com/36449837/215292404-e958b7e2-8ba2-47f1-9e96-c847c78e4dd4.png)
![image](https://user-images.githubusercontent.com/36449837/215292467-8a211735-b39f-4224-88be-0682658e29e8.png)
![image](https://user-images.githubusercontent.com/36449837/215292404-e958b7e2-8ba2-47f1-9e96-c847c78e4dd4.png)

This is the log incase its of any use
[Legacinator.log](https://github.com/ViGEm/BthPS3/files/10528439/Legacinator.log)

**Reply by nefarius** [reply]
[Found and documented the solution](https://vigem.org/projects/BthPS3/Frequently-Asked-Questions/#how-to-fix-previous-version-found-on-reinstall)!

**Comment by nefarius** [comment]
OK IDK where people get this idea from but **stop using Driver Store Explorer to remove BthPS3** or exactly that will happen! 

Uninstall it **using the setup**, like literally any application since Windows 95, it baffles me that I even have to explain this...

For those fortunate souls who decided to trash their system: [Legacinator got you covered](https://github.com/nefarius/Legacinator/releases/tag/v1.2.48)!

**Reply by Soupdragon1** [reply]
Theres nothing to uninstall as it doesn't appear in Add/Remove Programs, right clicking BthPS3Setup_x64.msi and selecting "Uninstall" or "Repair" from the list pops up a messagebox with a big red cross and "This action is only valid for products that are currently installed".
![image](https://user-images.githubusercontent.com/36449837/215292361-115d9a58-808e-4355-b1f1-65f95ff7bce0.png)

**Reply by CedyRica** [reply]
I used legacinator and now I can't use Bluetooth anymore 😭

**Reply by Timoteee** [reply]
> I used legacinator and now I can't use Bluetooth anymore 😭
You have to reset your windows to fix this issues

**Comment by RACE4EVER** [comment]
I've recently tried this method. I've deleted .sys files with "bthps3psm" naming usinf IOBitUnlocker (is free) then for registry edits RegmagiK and Registry Workshop to clean "bthps3psm" entries then restarted couple of times. After that i've checked by Legacinator? it hasn't found any tracks of BthPS3.

Then before installing be sure to turn ON bluetooth device and to have Bluetooth Radio Host driver. Installed, rebooted, works fine, pairing two DS3 at the same time

**Comment by nefarius** [comment]
[Found and documented the solution](https://vigem.org/projects/BthPS3/Frequently-Asked-Questions/#how-to-fix-previous-version-found-on-reinstall)!
